llvm/lib/Target/AMDGPU/AMDGPUSwLowerLDS.cpp

```diff
@@ -0,0 +1,1335 @@
+      AsanInfo.Instructions.insert(cast<Instruction>(NewRMW));
+      RMW->replaceAllUsesWith(NewRMW);
+      RMW->eraseFromParent();
+    } else if (AtomicCmpXchgInst *XCHG = dyn_cast<AtomicCmpXchgInst>(Inst)) {
+      Value *XCHGPtrOperand = XCHG->getPointerOperand();
+      Value *Replacement = getTranslatedGlobalMemoryGEPOfLDSPointer(
+          LoadMallocPtr, XCHGPtrOperand);
+      Value *NewXCHG = IRB.CreateAtomicCmpXchg(
+          Replacement, XCHG->getCompareOperand(), XCHG->getNewValOperand(),
+          XCHG->getAlign(), XCHG->getSuccessOrdering(),
+          XCHG->getFailureOrdering(), XCHG->getSyncScopeID());
+      cast<AtomicCmpXchgInst>(NewXCHG)->setVolatile(XCHG->isVolatile());
+      AsanInfo.Instructions.insert(cast<Instruction>(NewXCHG));
+      XCHG->replaceAllUsesWith(NewXCHG);
+      XCHG->eraseFromParent();
+    } else
+      report_fatal_error("Unimplemented LDS lowering instruction");
+  }
+  return;
+}
+
+void AMDGPUSwLowerLDS::poisonRedzones(Function *Func, Value *MallocPtr) {
+  auto &LDSParams = FuncLDSAccessInfo.KernelToLDSParametersMap[Func];
+  Type *Int64Ty = IRB.getInt64Ty();
+  Type *VoidTy = IRB.getVoidTy();
+  FunctionCallee AsanPoisonRegion = M.getOrInsertFunction(
+      StringRef("__asan_poison_region"),
+      FunctionType::get(VoidTy, {Int64Ty, Int64Ty}, false));
+
+  auto RedzonesVec = LDSParams.RedzoneOffsetAndSizeVector;
+  size_t VecSize = RedzonesVec.size();
+  for (unsigned i = 0; i < VecSize; i++) {
+    auto &RedzonePair = RedzonesVec[i];
+    uint64_t RedzoneOffset = RedzonePair.first;
+    uint64_t RedzoneSize = RedzonePair.second;
+    Value *RedzoneAddrOffset = IRB.CreateInBoundsGEP(
+        IRB.getInt8Ty(), MallocPtr, {IRB.getInt64(RedzoneOffset)});
+    Value *RedzoneAddress = IRB.CreatePtrToInt(RedzoneAddrOffset, Int64Ty);
+    IRB.CreateCall(AsanPoisonRegion,
+                   {RedzoneAddress, IRB.getInt64(RedzoneSize)});
+  }
+}
+
+void AMDGPUSwLowerLDS::lowerKernelLDSAccesses(Function *Func,
+                                              DomTreeUpdater &DTU) {
+  LLVM_DEBUG(dbgs() << "Sw Lowering Kernel LDS for : " << Func->getName());
+  auto &LDSParams = FuncLDSAccessInfo.KernelToLDSParametersMap[Func];
+  auto &Ctx = M.getContext();
+  auto *PrevEntryBlock = &Func->getEntryBlock();
+  SetVector<Instruction *> LDSInstructions;
+  getLDSMemoryInstructions(Func, LDSInstructions);
+
+  // Create malloc block.
+  auto *MallocBlock = BasicBlock::Create(Ctx, "Malloc", Func, PrevEntryBlock);
+
+  // Create WIdBlock block which has instructions related to selection of
+  // {0,0,0} indiex work item in the work group.
+  auto *WIdBlock = BasicBlock::Create(Ctx, "WId", Func, MallocBlock);
+  IRB.SetInsertPoint(WIdBlock, WIdBlock->begin());
+  DebugLoc FirstDL =
+      getOrCreateDebugLoc(&*PrevEntryBlock->begin(), Func->getSubprogram());
+  IRB.SetCurrentDebugLocation(FirstDL);
+  auto *const WIdx =
+      IRB.CreateIntrinsic(Intrinsic::amdgcn_workitem_id_x, {}, {});
```

Comment:
You ought to be stripping amdgpu-no-workitem-id-x and co. from the function and all transitive callers